File: PYTHON/blockgroup.py
import sys

'''
In German folklore, a drude (German: Drude, pl. Druden[1]) is a kind of malevolent nocturnal spirit 
(an elf (Alp) or kobold or a hag) associated with nightmares, prevalent especially in Southern 
Germany. Druden were said to participate in the Wild Hunt and were considered a particular 
class of demon in Alfonso de Spina's hierarchy[citation needed]. The word also 
came to be used as a generic term for "witch" in the 16th century (Hans Sachs). -Wikipedia
'''

# BlockGroup class to represent drudes (+/- polyominos) and their methods.

class BlockGroup:
	oriented = True # is this oriented
	sub = False # is this subtractive
	pairs = {} # set of blocks
	n = 0 # number of blocks

	bottomleft = [0, 0]
	topright = [0, 0]
	boundingbox = (0, 0)

	BOUND = sys.maxsize
	NIL = (BOUND, BOUND)

	def __init__(self, orientation, subtractive, v): # oriented, sub, pairs
		self.oriented = orientation
		self.sub = subtractive
		self.pairs = set(v) # directly create the set of pairs from the list of pairs (tuples)
		self.n = len(self.pairs)

		self.bottomleft = [BlockGroup.BOUND, BlockGroup.BOUND]
		self.topright = [-1 * BlockGroup.BOUND, -1 * BlockGroup.BOUND]

		for pair in self.pairs:
			self.bottomleft[0] = min(self.bottomleft[0], pair[0])
			self.bottomleft[1] = min(self.bottomleft[1], pair[1])
			self.topright[0] = max(self.topright[0], pair[0])
			self.topright[1] = max(self.topright[1], pair[1])

		self.boundingbox = (self.topright[0] - self.bottomleft[0] + 1, self.topright[1] - self.bottomleft[1] + 1)
	
	def updateBounds(self):
		self.topright = [-1 * BlockGroup.BOUND, -1 * BlockGroup.BOUND]
		self.bottomleft = [BlockGroup.BOUND, BlockGroup.BOUND]

		for pair in self.pairs:
			self.bottomleft[0] = min(self.bottomleft[0], pair[0])
			self.bottomleft[1] = min(self.bottomleft[1], pair[1])
			self.topright[0] = max(self.topright[0], pair[0])
			self.topright[1] = max(self.topright[1], pair[1])

		self.boundingbox = (self.topright[0] - self.bottomleft[0] + 1, self.topright[1] - self.bottomleft[1] + 1)
	 
	# Utility Functions
	
	def contains(self, pair):
	    return pair in self.pairs
	
	def add(self, pair):
	    if self.contains(pair):
	        return
	    self.n = self.n + 1
	    self.pairs.add(pair)
	
	def remove(self, pair):
	    self.pairs.discard(pair)
	    self.n = self.n - 1
	    
	def reset(self, v):
		self.pairs = set(v) # directly create the set of pairs from the list of pairs (tuples)
		self.n = len(self.pairs)

		self.bottomleft = [BlockGroup.BOUND, BlockGroup.BOUND]
		self.topright = [-1 * BlockGroup.BOUND, -1 * BlockGroup.BOUND]

		for pair in self.pairs:
			self.bottomleft[0] = min(self.bottomleft[0], pair[0])
			self.bottomleft[1] = min(self.bottomleft[1], pair[1])
			self.topright[0] = max(self.topright[0], pair[0])
			self.topright[1] = max(self.topright[1], pair[1])

		self.boundingbox = (self.topright[0] - self.bottomleft[0] + 1, self.topright[1] - self.bottomleft[1] + 1)
	
	def clone(self):
	    newpoints = []
	    for pair in self.pairs:
	        newpoints.append(pair)
	    
	    return BlockGroup(self.oriented, self.sub, newpoints)
	
	# General Functions
	
	def rotate(self, x):
	    while x < 0:
	        x = x + 1000
	    x = x % 4
	    if x <= 0:
	        return
	    
	    self.rotate(x - 1)
	    
	    res = []
	    for pair in self.pairs:
	        res.append((-1 * pair[1], pair[0]))
	    
	    self.reset(res)
	
	def move(self, disp):
	    res = []
	    for pair in self.pairs:
	        res.append((pair[0] + disp[0], pair[1] + disp[1]))
	    
	    self.reset(res)
	
	def invmov(self, disp):
	    res = []
	    for pair in self.pairs:
	        res.append((pair[0] - disp[0], pair[1] - disp[1]))
	    
	    self.reset(res)
	
	def normalize(self):
	    self.invmov(self.bottomleft)
	
	def removeRegion(self, other):
	    for o in other.pairs:
	        self.remove(o)
	
	def addRegion(self, other):
	    for o in other.pairs:
	        self.add(o)
	
	# Region Testing
	
	def containsbb(self, b):
	    if self.boundingbox[0] < b.boundingbox[0]:
	        return False
	    if self.boundingbox[1] > b.boundingbox[1]:
	        return False
	    return True
	
	def directoverlay(self, b):
	    if self.n < b.n:
	        return False
	    for p in b.pairs:
	        if not self.contains(p):
	            return False
	    
	    return True
	
	def fixedoverlay(self, b):
	    res = []
	    if self.n < b.n:
	        return res
	    
	    dx = self.bottomleft[0] - b.bottomleft[0]
	    dy = self.bottomleft[1] - b.bottomleft[1]
	    
	    test = b.clone()
	    test.move((dx, dy))
	    
	    width = abs(self.boundingbox[0] - b.boundingbox[0])
	    height = abs(self.boundingbox[1] - b.boundingbox[1])
	    
	    for i in range(width + 1):
	        for j in range(height + 1):
	            if self.directoverlay(test):
	                res.append((i, j))
	            test.move((0, 1))
	        test.move((1, -1 * (height + 1)))
	    
	    return res
	
	def overlay(self, b):
	    res = [[], [], [], []]
	    res[0] = self.fixedoverlay(b)
	    
	    if not b.oriented:
	        test = b.clone()
	        test.normalize()
	        for i in range(3):
	            test.rotate(1)
	            test.normalize()
	            res[i + 1] = self.fixedoverlay(test)
	    
	    return res
	
	def dfsUtil(self, region, v, index):
	    if (index >= len(v)):
	        return region.n == 0
	    
	    group = v[index].clone()
	    group.normalize()
	    options = region.overlay(group)
	    
	    res = False
	    for i in range(4):
	        group.move(region.bottomleft)
	        for op in options[i]:
	            group.move(op)
	            region.removeRegion(group)
	            
	            res = self.dfsUtil(region, v, index + 1) or res
	            if (res):
	                return True
	                
	            group.addRegion(group)
	            group.invmov(op)
	        
	        group.rotate(1)
	        group.normalize()
	    
	    return res
	
	def solve(self, v):
	    return self.dfsUtil(self.clone(), v, 0)

File: JAVA/BlockGroup.java
import java.awt.*;
import java.util.ArrayList;
import java.util.List;
import java.util.TreeSet;

public class BlockGroup extends Entity {
    public boolean oriented;
    public boolean sub;

    public TreeSet<Pair> pairs;
    public int n;

    public Pair bottomleft;
    public Pair topright;
    public Pair boundingbox;

    public BlockGroup(boolean orientation, boolean subtractive, Pair[] v, Color c) {
        oriented = orientation;
        sub = subtractive;
        pairs = new TreeSet<Pair>();
        n = v.length;
        color = c;

        bottomleft = new Pair(Integer.MAX_VALUE, Integer.MAX_VALUE);
        topright = new Pair(Integer.MIN_VALUE, Integer.MIN_VALUE);

        for (int i = 0; i < n; i++) {
            pairs.add(v[i]);
            bottomleft.x = Math.min(bottomleft.x, v[i].x);
            bottomleft.y = Math.min(bottomleft.y, v[i].y);
            topright.x = Math.max(topright.x, v[i].x);
            topright.y = Math.max(topright.y, v[i].y);
        }

        boundingbox = new Pair(topright.x - bottomleft.x + 1, topright.y - bottomleft.y + 1);
    }

    public BlockGroup(boolean orientation, boolean subtractive, Pair[] v) {
        this(orientation, subtractive, v, Color.YELLOW);
    }

    public void updateBounds() {
        bottomleft = new Pair(Integer.MAX_VALUE, Integer.MAX_VALUE);
        topright = new Pair(Integer.MIN_VALUE, Integer.MIN_VALUE);

        for (Pair p : pairs) {
            bottomleft.x = Math.min(bottomleft.x, p.x);
            bottomleft.y = Math.min(bottomleft.y, p.y);
            topright.x = Math.max(topright.x, p.x);
            topright.y = Math.max(topright.y, p.y);
        }

        boundingbox = new Pair(topright.x - bottomleft.x + 1, topright.y - bottomleft.y + 1);
    }

    // Utility Functions

    public boolean contains(Pair p) {
        return pairs.contains(p);
    }

    public void add(Pair p) {
        if (contains(p)) return;
        pairs.add(p.clone());
        n++;
    }

    public void remove(Pair p) {
        if (!contains(p)) return;
        pairs.remove(p);
        n--;
    }

    public void reset(Pair[] v) {
        n = v.length;
        pairs.clear();
        bottomleft = new Pair(Integer.MAX_VALUE, Integer.MAX_VALUE);
        topright = new Pair(Integer.MIN_VALUE, Integer.MIN_VALUE);

        for (int i = 0; i < n; i++) {
            pairs.add(v[i]);
            bottomleft.x = Math.min(bottomleft.x, v[i].x);
            bottomleft.y = Math.min(bottomleft.y, v[i].y);
            topright.x = Math.max(topright.x, v[i].x);
            topright.y = Math.max(topright.y, v[i].y);
        }

        boundingbox = new Pair(topright.x - bottomleft.x + 1, topright.y - bottomleft.y + 1);
    }

    public BlockGroup clone() {
        Pair[] v = new Pair[pairs.size()];
        int index = 0;
        for (Pair p : pairs) v[index++] = p.clone();
        return new BlockGroup(this.oriented, this.sub, v, this.color);
    }

    // General Functions

    public void rotate(int x) {
        while (x < 0) x += 8;
        x %= 4;
        if (x <= 0) return;
        rotate(x - 1);

        Pair[] res = new Pair[pairs.size()];
        int index = 0;
        for (Pair p : pairs) res[index++] = new Pair(-1 * p.y, p.x);
        reset(res);
    }

    public void move(Pair n) {
        Pair[] res = new Pair[pairs.size()];
        int index = 0;
        for (Pair p : pairs) res[index++] = p.add(n);
        reset(res);
    }

    public void invmov(Pair n) {
        move(n.inv());
    }

    public void normalize() {
        invmov(bottomleft);
    }

    public void removeRegion(BlockGroup o) {
        for (Pair p : o.pairs) this.remove(p);
    }

    public void addRegion(BlockGroup o) {
        for (Pair p : o.pairs) this.add(p);
    }

    public String toString() {
        StringBuilder sb = new StringBuilder("[O = ");
        sb.append(this.oriented ? "1" : "0");
        sb.append(" S = ");
        sb.append(this.sub ? "1" : "0");
        sb.append(" N = ");
        sb.append(Integer.toString(n));
        sb.append("] = [");
        for (Pair i : pairs) sb.append(i.toString());
        sb.append("]");
        return sb.toString();
    }

    // RegionTesting

    public boolean containsBB(BlockGroup b) { // Can the bounding box contain that belonging to region b?
        if (boundingbox.x < b.boundingbox.x) return false;
        if (boundingbox.y < b.boundingbox.y) return false;
        return true;
    }

    public boolean directoverlay(BlockGroup b) { // Does this region contain region b in terms of absolute coordinates?
        if (n < b.n) return false;
        for (Pair p : b.pairs) {
            if (!contains(p)) return false;
        }
        return true;
    }

    ArrayList<Pair> fixedoverlay(BlockGroup b) { // Does this region contain region b? Return the offset if yes, INT_MIN if no.
        ArrayList<Pair> res = new ArrayList<Pair>();
        if (n < b.n) return res;

        int dx = bottomleft.x - b.bottomleft.x;
        int dy = bottomleft.y - b.bottomleft.y;

        BlockGroup test = b.clone();
        test.move(new Pair(dx, dy));

        int width = Math.abs(boundingbox.x - b.boundingbox.x);
        int height = Math.abs(boundingbox.y - b.boundingbox.y);

        Pair carriagereturn = new Pair(1, -1 * height - 1);

        for (int i = 0; i <= width; i++) {
            for (int j = 0; j <= height; j++) {
                if (directoverlay(test)) res.add(new Pair(i, j));
                test.move(Pair.PLUS_Y);
            }
            test.move(carriagereturn);
        }

        return res;
    }

    ArrayList<ArrayList<Pair>> overlay(BlockGroup b) {
        ArrayList<ArrayList<Pair>> res = new ArrayList<ArrayList<Pair>>();
        res.add(fixedoverlay(b));
        if (!b.oriented) {
            BlockGroup test = b.clone();
            test.normalize();
            for (int i = 0; i < 3; i++) {
                test.rotate(1);
                test.normalize();
                res.add(fixedoverlay(test));
            }
        }

        return res;
    }

    // Now for the real thing

    public boolean dfsUtil(BlockGroup region, BlockGroup[] v, int index) {
        if (index >= v.length) return region.n == 0;

        BlockGroup group = v[index].clone();
        group.normalize();
        ArrayList<ArrayList<Pair>> options = region.overlay(group);

        boolean res = false;
        for (int i = 0; i < 4; i++) {
            group.move(region.bottomleft);
            for (Pair op : options.get(i)) {
                group.move(op);
                region.removeRegion(group);
                res |= dfsUtil(region, v, index + 1);
                if (res) return true;
                region.addRegion(group);
                group.invmov(op);
            }
            if (group.oriented) break;
            group.rotate(1);
            group.normalize();
        }
        return res;
    }

    public boolean solve(BlockGroup[] v) {
        int diff = n;
        for (BlockGroup i : v) {
            if (i.sub) diff += i.n;
            else diff -= i.n;
        }
        if (diff != 0) return false;
        return dfsUtil(clone(), v, 0);
    }
}
